game/plugin/skill/firemaking/src/firemaking.plugin.kts

```diff
@@ -0,0 +1,138 @@
+                groundLog = GroundItem.dropped(player.world, player.position, Item(log.id), player)
+                player.world.spawn(groundLog)
+            } else {
+                player.sendMessage("You cannot light a fire here.")
+                stop()
+                return
+            }
+            started = true
+        }
+
+        //light the fire
+        player.playAnimation(LIGHT_ANIMATION)
+
+        if (successfulLight()) {
+            if (lightFire(Direction.WEST) ||
+                    lightFire(Direction.EAST) ||
+                    lightFire(Direction.NORTH) ||
+                    lightFire(Direction.SOUTH) ||
+                    lightFire(Direction.NONE)
+                    ) {
+                player.sendMessage("The fire catches and the logs begin to burn.")
+                player.skillSet.addExperience(Skill.FIREMAKING, log.xp)
+            } else {
+                player.sendMessage("You cannot light a fire here.")
+            }
+            stop()
+        }
+    }
+
+    override fun stop() {
+        super.stop()
+        player.stopAnimation()
+    }
+
+    fun successfulLight(): Boolean {
+        //TODO: This is from lare96, as he mentioned, we need to find the actual chance
+        val playerLevel = player.skillSet.getSkill(Skill.FIREMAKING).currentLevel
+        val lowChance = playerLevel - log.level + 5
+        if (lowChance > 30) {
+            return 30 > rand.nextInt(40)
+        } else {
+            return lowChance > rand.nextInt(40)
+        }
+    }
+
+
+    fun lightFire(direction: Direction): Boolean {
+        if (canLight(direction)) {
+            val fire = DynamicGameObject.createPublic(player.world, FIRE_OBJ, player.position, 10, 0)
+            player.walkingQueue.addFirstStep(player.position.step(1, direction))
+            val region = player.world.regionRepository.fromPosition(player.position)
+            region.removeEntity(groundLog)
+            player.world.spawn(fire)
+
+            //TODO: burn time = log level * 5 + rand(30). I think this is right can someone verify?
+
+            val burnTime = (log.level * 5 + rand.nextInt(30) * 1000) / 600 //convert to pulses
+            player.world.schedule(object: ScheduledTask(burnTime, false) {
+                override fun execute() {
+                    region.removeEntity(fire)
+                    player.world.spawn(GroundItem.create(player.world, fire.position, Item(ASH)))
+                    this.stop()
+                }
+
+            })
+            return true
+        }
+        return false
+    }
+
+    fun canLight(direction: Direction): Boolean {
+        val region = player.world.regionRepository.fromPosition(player.position)
+        if (direction == Direction.NONE) {
+            return true
+        }
+        if (region.traversable(player.position, EntityType.PLAYER, direction)) {
+            return true
+        }
+        return false
+    }
+
+}
+
+
+on { ItemOnItemMessage::class }
```

Comment:
@Major- didn't you write something in Ruby to handle these ItemOnItem pairs?

Reply:
I removed it because it wasn't very good, maybe we could look into something for kotlin

Reply:
Is the ItemOnItemMessage not the correct way to do this?

Reply:
Nope this is right, we're just discussing how we can make it a bit nicer

Reply:
ah, ok